File: beautyhq_flutter/lib/core/models/transaction.dart
import 'package:equatable/equatable.dart';
import 'package:flutter/material.dart';

enum PaymentMethod {
  cash,
  card,
  giftCard,
  loyalty,
  other,
}

extension PaymentMethodExtension on PaymentMethod {
  String get displayName {
    switch (this) {
      case PaymentMethod.cash:
        return 'Cash';
      case PaymentMethod.card:
        return 'Card';
      case PaymentMethod.giftCard:
        return 'Gift Card';
      case PaymentMethod.loyalty:
        return 'Loyalty Points';
      case PaymentMethod.other:
        return 'Other';
    }
  }

  IconData get icon {
    switch (this) {
      case PaymentMethod.cash:
        return Icons.payments;
      case PaymentMethod.card:
        return Icons.credit_card;
      case PaymentMethod.giftCard:
        return Icons.card_giftcard;
      case PaymentMethod.loyalty:
        return Icons.stars;
      case PaymentMethod.other:
        return Icons.more_horiz;
    }
  }

  static PaymentMethod fromString(String value) {
    switch (value.toLowerCase()) {
      case 'cash':
        return PaymentMethod.cash;
      case 'card':
        return PaymentMethod.card;
      case 'gift_card':
      case 'giftcard':
        return PaymentMethod.giftCard;
      case 'loyalty':
        return PaymentMethod.loyalty;
      default:
        return PaymentMethod.other;
    }
  }
}

enum TransactionStatus {
  pending,
  completed,
  refunded,
  failed,
}

extension TransactionStatusExtension on TransactionStatus {
  String get displayName {
    switch (this) {
      case TransactionStatus.pending:
        return 'Pending';
      case TransactionStatus.completed:
        return 'Completed';
      case TransactionStatus.refunded:
        return 'Refunded';
      case TransactionStatus.failed:
        return 'Failed';
    }
  }

  Color get color {
    switch (this) {
      case TransactionStatus.pending:
        return Colors.orange;
      case TransactionStatus.completed:
        return Colors.green;
      case TransactionStatus.refunded:
        return Colors.blue;
      case TransactionStatus.failed:
        return Colors.red;
    }
  }

  static TransactionStatus fromString(String value) {
    switch (value.toLowerCase()) {
      case 'pending':
        return TransactionStatus.pending;
      case 'completed':
        return TransactionStatus.completed;
      case 'refunded':
        return TransactionStatus.refunded;
      case 'failed':
        return TransactionStatus.failed;
      default:
        return TransactionStatus.pending;
    }
  }
}

class Transaction extends Equatable {
  final String id;
  final String businessId;
  final String? clientId;
  final String? clientName;
  final String? appointmentId;
  final double amount;
  final double? tip;
  final double? discount;
  final double total;
  final PaymentMethod paymentMethod;
  final TransactionStatus status;
  final String? notes;
  final List<TransactionItem>? items;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Transaction({
    required this.id,
    required this.businessId,
    this.clientId,
    this.clientName,
    this.appointmentId,
    required this.amount,
    this.tip,
    this.discount,
    required this.total,
    required this.paymentMethod,
    required this.status,
    this.notes,
    this.items,
    required this.createdAt,
    required this.updatedAt,
  });

  String get formattedTotal => '\$${total.toStringAsFixed(2)}';

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'] as String,
      businessId: json['businessId'] as String,
      clientId: json['clientId'] as String?,
      clientName: json['clientName'] as String?,
      appointmentId: json['appointmentId'] as String?,
      amount: (json['amount'] as num).toDouble(),
      tip: (json['tip'] as num?)?.toDouble(),
      discount: (json['discount'] as num?)?.toDouble(),
      total: (json['total'] as num).toDouble(),
      paymentMethod:
          PaymentMethodExtension.fromString(json['paymentMethod'] as String),
      status:
          TransactionStatusExtension.fromString(json['status'] as String),
      notes: json['notes'] as String?,
      items: json['items'] != null
          ? (json['items'] as List)
              .map((i) => TransactionItem.fromJson(i as Map<String, dynamic>))
              .toList()
          : null,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'businessId': businessId,
      'clientId': clientId,
      'clientName': clientName,
      'appointmentId': appointmentId,
      'amount': amount,
      'tip': tip,
      'discount': discount,
      'total': total,
      'paymentMethod': paymentMethod.name,
      'status': status.name,
      'notes': notes,
      'items': items?.map((i) => i.toJson()).toList(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  @override
  List<Object?> get props => [
        id,
        businessId,
        clientId,
        clientName,
        appointmentId,
        amount,
        tip,
        discount,
        total,
        paymentMethod,
        status,
        notes,
        items,
        createdAt,
        updatedAt,
      ];
}

class TransactionItem extends Equatable {
  final String id;
  final String name;
  final int quantity;
  final double price;
  final double total;

  const TransactionItem({
    required this.id,
    required this.name,
    required this.quantity,
    required this.price,
    required this.total,
  });

  factory TransactionItem.fromJson(Map<String, dynamic> json) {
    return TransactionItem(
      id: json['id'] as String,
      name: json['name'] as String,
      quantity: json['quantity'] as int,
      price: (json['price'] as num).toDouble(),
      total: (json['total'] as num).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'quantity': quantity,
      'price': price,
      'total': total,
    };
  }

  @override
  List<Object?> get props => [id, name, quantity, price, total];
}

File: BeautyHQ-iOS/BeautyHQ/Models/Transaction.swift
import Foundation
import SwiftUI

// Line item from backend
struct TransactionLineItem: Codable, Identifiable {
    var id: String { "\(type)-\(name)" }
    let type: String
    let name: String
    let quantity: Int
    let unitPrice: Double
    let totalPrice: Double

    // Custom decoder for Prisma Decimal fields
    enum CodingKeys: String, CodingKey {
        case type, name, quantity, unitPrice, totalPrice
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        type = try container.decode(String.self, forKey: .type)
        name = try container.decode(String.self, forKey: .name)
        quantity = try container.decode(Int.self, forKey: .quantity)
        unitPrice = try container.decodeFlexibleDouble(forKey: .unitPrice)
        totalPrice = try container.decodeFlexibleDouble(forKey: .totalPrice)
    }
}

struct Transaction: Codable, Identifiable {
    let id: String
    let transactionNumber: String?
    let date: Date?
    let createdAt: Date?
    let clientName: String?
    let staffName: String?
    let subtotal: Double?
    let taxAmount: Double?
    let tipAmount: Double?
    let discountAmount: Double?
    let totalAmount: Double?
    let paymentMethod: String?
    let status: TransactionStatus?
    let type: TransactionType?
    let lineItems: [TransactionLineItem]?

    // Custom decoder for Prisma Decimal fields
    enum CodingKeys: String, CodingKey {
        case id, transactionNumber, date, createdAt, clientName, staffName
        case subtotal, taxAmount, tipAmount, discountAmount, totalAmount
        case paymentMethod, status, type, lineItems
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        transactionNumber = try container.decodeIfPresent(String.self, forKey: .transactionNumber)
        date = try container.decodeIfPresent(Date.self, forKey: .date)
        createdAt = try container.decodeIfPresent(Date.self, forKey: .createdAt)
        clientName = try container.decodeIfPresent(String.self, forKey: .clientName)
        staffName = try container.decodeIfPresent(String.self, forKey: .staffName)
        subtotal = try container.decodeFlexibleDoubleIfPresent(forKey: .subtotal)
        taxAmount = try container.decodeFlexibleDoubleIfPresent(forKey: .taxAmount)
        tipAmount = try container.decodeFlexibleDoubleIfPresent(forKey: .tipAmount)
        discountAmount = try container.decodeFlexibleDoubleIfPresent(forKey: .discountAmount)
        totalAmount = try container.decodeFlexibleDoubleIfPresent(forKey: .totalAmount)
        paymentMethod = try container.decodeIfPresent(String.self, forKey: .paymentMethod)
        status = try container.decodeIfPresent(TransactionStatus.self, forKey: .status)
        type = try container.decodeIfPresent(TransactionType.self, forKey: .type)
        lineItems = try container.decodeIfPresent([TransactionLineItem].self, forKey: .lineItems)
    }

    // Convenience computed properties
    var total: Double { totalAmount ?? 0 }
    var tax: Double { taxAmount ?? 0 }
    var tip: Double { tipAmount ?? 0 }
    var discount: Double { discountAmount ?? 0 }

    var formattedTotal: String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = "USD"
        return formatter.string(from: NSNumber(value: total)) ?? "$\(total)"
    }

    var paymentMethodDisplay: String {
        paymentMethod ?? "Unknown"
    }
}

enum TransactionType: String, Codable, CaseIterable {
    case sale = "SALE"
    case refund = "REFUND"
    case void = "VOID"
    case adjustment = "ADJUSTMENT"

    var displayName: String {
        rawValue.capitalized
    }

    var color: Color {
        switch self {
        case .sale: return .green
        case .refund: return .orange
        case .void: return .red
        case .adjustment: return .gray
        }
    }

    var icon: String {
        switch self {
        case .sale: return "cart.fill"
        case .refund: return "arrow.uturn.backward"
        case .void: return "xmark.circle.fill"
        case .adjustment: return "arrow.left.arrow.right"
        }
    }
}

enum TransactionStatus: String, Codable {
    case pending = "PENDING"
    case completed = "COMPLETED"
    case failed = "FAILED"
    case refunded = "REFUNDED"
}

enum PaymentMethod: String, Codable, CaseIterable {
    case cash = "CASH"
    case card = "CARD"
    case applePay = "APPLE_PAY"
    case googlePay = "GOOGLE_PAY"
    case giftCard = "GIFT_CARD"
    case points = "POINTS"
    case split = "SPLIT"

    var displayName: String {
        switch self {
        case .cash: return "Cash"
        case .card: return "Card"
        case .applePay: return "Apple Pay"
        case .googlePay: return "Google Pay"
        case .giftCard: return "Gift Card"
        case .points: return "Points"
        case .split: return "Split"
        }
    }

    var icon: String {
        switch self {
        case .cash: return "banknote.fill"
        case .card: return "creditcard.fill"
        case .applePay: return "apple.logo"
        case .googlePay: return "g.circle.fill"
        case .giftCard: return "giftcard.fill"
        case .points: return "star.fill"
        case .split: return "rectangle.split.2x1.fill"
        }
    }
}

// MARK: - Create Transaction
struct CreateTransactionRequest: Codable {
    let clientId: String?
    let staffId: String?
    let appointmentId: String?
    let items: [CreateTransactionItem]
    let paymentMethod: PaymentMethod
    let tip: Double?
    let discount: Double?
    let notes: String?
}

enum TransactionItemType: String, Codable {
    case service = "SERVICE"
    case product = "PRODUCT"
    case giftCard = "GIFT_CARD"
}

struct CreateTransactionItem: Codable {
    let type: TransactionItemType
    let itemId: String
    let quantity: Int
    let unitPrice: Double
    let discount: Double?
}
